File: tools_box/tools_box/report/sales_by_product_report/sales_by_product_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Sales by Product Report"] = {
	"filters": [
        {
                        fieldname: "from_date",
                        label: __("From Date"),
                        fieldtype: "Date",
                },
        {
                        fieldname: "to_date",
                        label: __("To Date"),
                        fieldtype: "Date",
                },
        {
                        fieldname: "item",
                        label: __("Item"),
                        fieldtype: "Link",
                        options: "Item",
                },
        {
                        fieldname: "customer",
                        label: __("Customer"),
                        fieldtype: "Link",
                        options: "Customer",
                },
        {
                        fieldname: "territory",
                        label: __("Territory"),
                        fieldtype: "Link",
                        options: "Territory",
                },
	]
}

File: tools_box/tools_box/report/sales_by_product_report/sales_by_product_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

def execute(filters=None):
	if not filters: filters ={}
	columns = ["Item:Link/Item:300","Invoice Date:Date:110","Sales Invoice::100","Customer:Link/Customer:200","Qty:Float:50","Amount:Currency:100","Territory:Link/Territory:100"]
	conditions = ""
	if filters.get("from_date"):
		conditions += " AND si.posting_date >= %(from_date)s"
	if filters.get("to_date"):
		conditions += " AND si.posting_date <= %(to_date)s"
	if filters.get("customer"):
		conditions += " AND si.customer = %(customer)s"
	if filters.get("item"):
		conditions += " AND sii.item_code = %(item)s"
	if filters.get("territory"):
		conditions += " AND c.territory = %(territory)s"

	data = frappe.db.sql("""SELECT CONCAT(sii.item_name,' (',sii.item_code,')'),si.posting_date,si.name,si.customer,sii.qty,sii.amount,c.territory FROM `tabSales Invoice` si, `tabSales Invoice Item` sii, `tabCustomer` c WHERE si.name = sii.parent AND c.name = si.customer %s ORDER BY sii.item_name,si.posting_date""" % conditions,filters)
	return columns, data
